hudi-common/src/main/java/org/apache/hudi/common/table/read/KeyBasedFileGroupRecordBuffer.java

```diff
@@ -121,6 +129,18 @@ public boolean containsLogRecord(String recordKey) {
 
   protected boolean hasNextBaseRecord(T baseRecord) throws IOException {
     String recordKey = readerContext.getRecordKey(baseRecord, readerSchema);
+    int comparison;
+    while (sortOutput && !getLogRecordKeysSorted().isEmpty() && (comparison = getLogRecordKeysSorted().peek().compareTo(recordKey)) <= 0) {
```

Comment:
either we have a separate record buffer impl (by sub-class `KeyBasedFileGroupRecordBuffer`) or move the sorting into `FileGroupReaderBasedAppendHandle` itself.